app/(dashboard)/dashboard/interviews/_actions/export.ts

```diff
@@ -0,0 +1,147 @@
+'use server';
+
+import FileExportManager from '~/lib/network-exporters/FileExportManager';
+import { api } from '~/trpc/server';
+import { getServerSession } from '~/utils/auth';
+import { formatExportableSessions, getRemoteProtocolID } from './utils';
+import { trackEvent } from '~/analytics/utils';
+import { type ExportOptions } from '../_components/ExportInterviewsDialog';
+
+type UploadData = {
+  key: string;
+  url: string;
+  name: string;
+  size: number;
+};
+
+type UpdateItems = {
+  statusText: string;
+  progress: number;
+};
+
+type FailResult = {
+  data: null;
+  error: unknown;
+  message: string;
+};
+
+type SuccessResult = {
+  data: UploadData;
+  error: null;
+  message: string;
+};
+
+export const exportSessions = async (
+  interviewIds: { id: string }[],
+  exportOptions?: ExportOptions,
+) => {
+  const session = await getServerSession();
+
+  if (!session) {
+    throw new Error('You must be logged in to export interview sessions!.');
+  }
+
+  const installedProtocols = await api.protocol.get.all.query();
+  const interviewsSessions =
+    await api.interview.get.manyByIds.query(interviewIds);
+
+  const fileExportManager = new FileExportManager(exportOptions);
+
+  fileExportManager.on('begin', () => {
+    // eslint-disable-next-line no-console
+    console.log({
+      statusText: 'Starting export...',
+      percentProgress: 0,
+    });
+  });
+
+  fileExportManager.on('update', ({ statusText, progress }: UpdateItems) => {
+    // eslint-disable-next-line no-console
+    console.log({
+      statusText,
+      percentProgress: progress,
+    });
+  });
+
+  fileExportManager.on('session-exported', (sessionId: unknown) => {
+    if (!sessionId || typeof sessionId !== 'string') {
+      // eslint-disable-next-line no-console
+      console.warn('session-exported event did not contain a sessionID');
+      return;
+    }
+    // eslint-disable-next-line no-console
+    console.log('session-exported success sessionId:', sessionId);
+  });
+
+  fileExportManager.on('error', (errResult: FailResult) => {
+    // eslint-disable-next-line no-console
+    console.log('Session export failed, Error:', errResult.message);
+  });
+
+  fileExportManager.on('finished', ({ statusText, progress }: UpdateItems) => {
+    // eslint-disable-next-line no-console
+    console.log({
+      statusText,
+      percentProgress: progress,
+    });
+  });
+
+  const formattedSessions = formatExportableSessions(
+    interviewsSessions,
+    installedProtocols,
+  );
```

Comment:
This function should be refactored:

1. No need to pass in `installedProtocols` - the Protocol should be included in the Sessions query.
2. This processing should happen inside the tRPC endpoint that returns the sessions

Reply:
Done ✅, refactored the function but using it inside the server action because of the tRPC endpoint's cashing behavior.